pkg/apis/batch/validation/validation.go

```diff
@@ -168,6 +170,58 @@ func validateJobSpec(spec *batch.JobSpec, fldPath *field.Path, opts apivalidatio
 		}
 	}
 
+	if spec.PodFailurePolicy != nil {
+		containerNames := sets.NewString()
+		for _, containerSpec := range spec.Template.Spec.Containers {
+			containerNames.Insert(containerSpec.Name)
+		}
+		for _, containerSpec := range spec.Template.Spec.InitContainers {
+			containerNames.Insert(containerSpec.Name)
+		}
+		for i, rule := range spec.PodFailurePolicy.Rules {
+			rulePath := fldPath.Child("podFailurePolicy").Index(i)
+
+			if rule.Action != batch.PodFailurePolicyActionIgnore && rule.Action != batch.PodFailurePolicyActionFailJob && rule.Action != batch.PodFailurePolicyActionCount {
+				actionPath := rulePath.Child("action")
+				allErrs = append(allErrs, field.NotSupported(actionPath, rule.Action, []string{string(batch.PodFailurePolicyActionIgnore), string(batch.PodFailurePolicyActionFailJob), string(batch.PodFailurePolicyActionCount)}))
+			}
+			if rule.OnExitCodes != nil {
+				onExitCodesPath := rulePath.Child("onExitCodes")
+				if rule.OnExitCodes.Operator != batch.PodFailurePolicyOnExitCodesOpIn && rule.OnExitCodes.Operator != batch.PodFailurePolicyOnExitCodesOpNotIn {
+					path := onExitCodesPath.Child("operator")
+					allErrs = append(allErrs, field.NotSupported(path, rule.OnExitCodes.Operator, []string{string(batch.PodFailurePolicyOnExitCodesOpIn), string(batch.PodFailurePolicyOnExitCodesOpNotIn)}))
+				}
+				if rule.OnExitCodes.ContainerName != nil && !containerNames.Has(*rule.OnExitCodes.ContainerName) {
+					path := onExitCodesPath.Child("containerName")
+					allErrs = append(allErrs, field.Invalid(path, *rule.OnExitCodes.ContainerName, "must be one of the container names in the pod template"))
+				}
+				if rule.OnExitCodes.Operator == batch.PodFailurePolicyOnExitCodesOpIn {
+					for j, exitCodeValue := range rule.OnExitCodes.Values {
+						if exitCodeValue == 0 {
+							path := onExitCodesPath.Child("values").Index(j)
+							allErrs = append(allErrs, field.Invalid(path, exitCodeValue, "must not be 0 for the In operator"))
+						}
+					}
+				}
+			}
```

Comment:
I expected validation that the exit codes list:
* contained at least one item (required error otherwise)
* contained no duplicates
* *maybe* contains no negative numbers (check with sig-node if negative exit codes are ever possible)

Also, should we require it be sorted?

Reply:
Added validation for at least one item and no duplicates

As for negative exit codes, I don't think we have any additional validation for container exit codes, but maybe @bobbypage can comment?
As for sorting, do we have another example in the API where we allow to specify a list of integers? Maybe we could be consistent with the other place?

Reply:
> As for negative exit codes, I don't think we have any additional validation for container exit codes, but maybe @bobbypage can comment?

That seems more likely to have been an oversight than intentional. If there's any way we can legitimately get a negative exitCode, then I agree we shouldn't validate that here, but if not, let's limit this to `exitCode` values we could actually encounter.

> As for sorting, do we have another example in the API where we allow to specify a list of integers? Maybe we could be consistent with the other place?

These were what I found:

```go
SupplementalGroups []int64 // in PodSecurityContext, no duplicate or ordering validation is done

Ports []int32 // in PodPortForwardOptions, no validation is done

CompletedIndexes []string // in JobStatus, values must be non-negative and sorted
```

ensuring no duplicates and sorted seems ok to me

Reply:
Ok, I've added validation to ensure sorted and no duplicates.

Reply:
As for the negative exit codes, I was able to get a negative exit code on Windows 10 with this simple Python script:
```python
import sys
sys.exit(-12)
```
on Linux the script executed in bash returns 244 (256 - 12). However, it seems possible with some trickery too get negative exit codes on Linux too (haven't tried): https://unix.stackexchange.com/questions/418784/what-is-the-min-and-max-values-of-exit-codes-in-linux.

Given that we generally support Windows containers in Kubernetes it seems negative exit codes may happen. On the other hand, only values 0-255 seem to be sane :). 

Maybe @krmayankk or @dchen1107 could comment as node sig-node reviewers?

Reply:
if there's any question that we might be able to get negative exit codes, leaving that possible seems fine

Reply:
Yes, in CRI api it seems that exitCode is defined as `int32`, so it would be possible for the runtime to report back a negative exit code - https://github.com/kubernetes/cri-api/blob/master/pkg/apis/runtime/v1alpha2/api.proto#L1131-L1132 . Now if that actually happens on linux is a different story (since I believe it should wrap around...)

Though since you mentioned it is possible on Windows it makes sense to cover that case.